src/vdaf/mastic.rs

```diff
@@ -379,32 +382,418 @@ where
+                (
+                    MasticPrepareShare {
+                        vidpf_proof: eval_proof.into_seed(),
+                        szk_query_share_opt: None,
+                    },
+                    MasticPrepareState {
+                        output_shares: MasticOutputShare::<T::Field>::from(output_shares),
+                        szk_query_state: None,
+                        verifier_len: None,
+                    },
+                )
+            };
+        Ok((prep_state, prep_share))
+    }
+
+    fn prepare_shares_to_prepare_message<
+        M: IntoIterator<Item = MasticPrepareShare<T::Field, SEED_SIZE>>,
+    >(
+        &self,
+        _agg_param: &MasticAggregationParam,
+        inputs: M,
+    ) -> Result<MasticPrepareMessage<SEED_SIZE>, VdafError> {
+        let mut inputs_iter = inputs.into_iter();
+        let leader_share = inputs_iter.next().ok_or(VdafError::Uncategorized(
+            "No leader share received".to_string(),
+        ))?;
+        let helper_share = inputs_iter.next().ok_or(VdafError::Uncategorized(
+            "No helper share received".to_string(),
+        ))?;
+        if inputs_iter.next().is_some() {
+            return Err(VdafError::Uncategorized(
+                "Received more than two prepare shares".to_string(),
+            ));
+        };
+        if leader_share.vidpf_proof != helper_share.vidpf_proof {
+            return Err(VdafError::Uncategorized(
+                "Vidpf proof verification failed".to_string(),
+            ));
+        };
+        match (
+            leader_share.szk_query_share_opt,
+            helper_share.szk_query_share_opt,
+        ) {
+            (Some(leader_query_share), Some(helper_query_share)) => Ok(self
+                .szk
+                .merge_verifiers(leader_query_share, helper_query_share)?),
+            (None, None) => Ok(None),
+            (_, _) => Err(VdafError::Uncategorized(
+                "Only one of leader and helper query shares is present".to_string(),
+            )),
+        }
+    }
+
+    fn prepare_next(
+        &self,
+        state: MasticPrepareState<T::Field, SEED_SIZE>,
+        input: MasticPrepareMessage<SEED_SIZE>,
+    ) -> Result<PrepareTransition<Self, SEED_SIZE, NONCE_SIZE>, VdafError> {
+        match (state, input) {
+            (
+                MasticPrepareState {
+                    output_shares,
+                    szk_query_state: _,
+                    verifier_len: _,
+                },
+                None,
+            ) => Ok(PrepareTransition::Finish(output_shares)),
+            (
+                MasticPrepareState {
+                    output_shares,
+                    szk_query_state,
+                    verifier_len: _,
+                },
+                Some(ref joint_share),
+            ) => {
+                self.szk.decide(szk_query_state, joint_share)?;
+                Ok(PrepareTransition::Finish(output_shares))
+            }
+        }
+    }
```

Comment:
* Avoiding the match statement here will massively simplify control flow.
* Also, a cool Rust tip: You can unpack a struct in the parameters of the function!
```suggestion
    fn prepare_next(
        &self,
        MasticPrepareState {
            output_shares,
            szk_query_state,
            verifier_len: _,
        }: MasticPrepareState<T::Field, SEED_SIZE>,
        input: MasticPrepareMessage<SEED_SIZE>,
    ) -> Result<PrepareTransition<Self, SEED_SIZE, NONCE_SIZE>, VdafError> {
        if let Some(ref joint_share) = input {
            self.szk.decide(szk_query_state, joint_share)?;
        }

        Ok(PrepareTransition::Finish(output_shares))
    }
```